README.md

```diff
@@ -1,19 +1,122 @@
-
-Here is an example of computing the gradient of a function that goes from `R^3` to `R^2`:
-
-https://github.com/Mikolaj/horde-ad/blob/40951dab17ae90c5ce528a59216a39a3727eccb5/test/common/TestSingleGradient.hs#L442-L516
-
-Elsewhere in the same file is a computation of the forward derivative of the function and a QuickCheck test relating it to the gradient. It uses the same definition of the objective function and the same glue code for grouping parameters, etc. The ratio of signal to noise (maths to glue code) is much higher in more complex functions, e.g., neural networks. Several sample neural networks for MNIST digit classification are gathered in
+# HORDE-AD: Higher Order Reverse Derivatives Efficiently 
+
+This is an Automatic Differentiation library based on the paper _"Provably correct, asymptotically efficient, higher-order reverse-mode automatic differentiation"_ by Krawiec, Krishnaswami, Peyton Jones, Ellis, Fitzgibbon, and Eisenberg.
+
+This is an early prototype, both in terms of the engine performance and the API and toolbox for the library user. The user should be ready to add missing primitives, as well as obvious tools that should be predefined but aren't. One can already build with the library all basic neural network architectures, such as fully connected, recurrent, convolutional and residual networks. The library, due to its loose coupling of differentiation and data containers, can naturally handle exotic variants of such networks that wouldn't express well in a language of matrices or tensors. 
+
+Applications outside machine learning are very much in scope, given that the notion of a neural network is not hardwired into the formalism, but ad hoc redefined from basic blocks of general automatic differentiation whenever it's needed.
+
+Mature Haskell libraries with similar capabilities, but varying efficiency, are https://hackage.haskell.org/package/ad and https://hackage.haskell.org/package/backprop. See also https://github.com/Mikolaj/horde-ad/blob/master/CREDITS.md.  
+The benchmarks at TODO show that this library has performance highly competitive with (i.e. faster than) those and PyTorch on CPU.  
+It is hoped that the separation of AD logic from matrix and tensor manipulation (deferred to [hmatrix] and [orthotope]) will enable similar speedups on numerical accelerators.
+
+## Computing the derivative of a simple function
+
+Here is an example of computing the gradient of a function that goes from `R^3` to `R`:
+
+```Haskell
+-- A function that goes from `R^3` to `R`. 
+-- [Arguments are tupled, could just as easily be curried. -awf]
```

Comment:
These [-awf] comments are just for the pull request, so I didn't forget to make them.  They would be removed before merging.